bin/light-base/src/database.rs

```diff
@@ -0,0 +1,182 @@
+    chain,
+    database::finalized_serialize,
+    libp2p::{multiaddr, PeerId},
+};
+
+use crate::{network_service, platform, sync_service};
+
+/// A decoded database.
+pub struct DatabaseContent {
+    /// Hash of the genesis block, as provided to [`encode_database`].
+    pub genesis_block_hash: [u8; 32],
+    /// Information about the finalized chain.
+    pub chain_information: chain::chain_information::ValidChainInformation,
+    /// List of nodes that were known to be part of the peer-to-peer network when the database
+    /// was encoded.
+    pub known_nodes: Vec<(PeerId, Vec<multiaddr::Multiaddr>)>,
+}
+
+/// Serializes the finalized state of the chain, using the given services.
+///
+/// The returned string is guaranteed to not exceed `max_size` bytes. A truncated or invalid
+/// database is intentionally returned if `max_size` is too low to fit all the information.
+pub async fn encode_database<TPlat: platform::Platform>(
+    network_service: &network_service::NetworkService<TPlat>,
+    sync_service: &sync_service::SyncService<TPlat>,
+    genesis_block_hash: &[u8; 32],
+    max_size: usize,
+) -> String {
+    // Craft the structure containing all the data that we would like to include.
+    let mut database_draft = SerdeDatabase {
+        genesis_hash: hex::encode(genesis_block_hash),
+        chain: match sync_service.serialize_chain_information().await {
+            Some(ci) => {
+                let encoded =
+                    finalized_serialize::encode_chain(&ci, sync_service.block_number_bytes());
+                serde_json::from_str(&encoded).unwrap()
+            }
+            None => {
+                // If the chain information can't be obtained, we just return a dummy value that
+                // will intentionally fail to decode if passed back.
+                let dummy_message = "<unknown>";
+                return if dummy_message.len() > max_size {
+                    String::new()
+                } else {
+                    dummy_message.to_owned()
+                };
+            }
+        },
+        nodes: network_service
+            .discovered_nodes(0) // TODO: hacky chain_index
+            .await
+            .map(|(peer_id, addrs)| {
+                (
+                    peer_id.to_base58(),
+                    addrs.map(|a| a.to_string()).collect::<Vec<_>>(),
+                )
+            })
+            .collect(),
+    };
+
+    // Cap the database length to the maximum size.
+    loop {
+        let serialized = serde_json::to_string(&database_draft).unwrap();
+        if serialized.len() <= max_size {
+            // Success!
+            return serialized;
+        }
+
+        if database_draft.nodes.is_empty() {
+            // Can't shrink the database anymore. Return the string `"<too-large>"` which will
+            // fail to decode but will indicate what is wrong.
+            let dummy_message = "<too-large>";
+            return if dummy_message.len() >= max_size {
```

Comment:
```suggestion
            return if dummy_message.len() > max_size {
```

to be consistent with the similar check above